src/metabase/lib/stage.cljc

```diff
@@ -346,12 +346,22 @@
     query
     (update query :stages pop)))
 
-(mu/defn drop-stage-if-empty :- ::lib.schema/query
-  "Drops the final stage in the pipeline IF the stage is empty of clauses, otherwise no-op"
+(defn- remove-stage
+  "Removes the stage at `stage-number` from `query`.
+  This might result in a broken query, so it's expected to used on empty stages."
+  [query stage-number]
+  (update query :stages (fn [stages]
+                          (let [[prefix suffix] (split-at stage-number stages)]
+                            (into (vec prefix) (next suffix))))))
+
+(mu/defn drop-empty-stages :- ::lib.schema/query
+  "Drops all empty stages in the pipeline."
   [query :- ::lib.schema/query]
-  (if-not (has-clauses? query -1)
-    (drop-stage query)
-    query))
+  (reduce (fn [query stage-number]
+            (cond-> query
+              (not (has-clauses? query stage-number)) (remove-stage stage-number)))
+          query
+          (range (lib.util/canonical-stage-index query -1) 0 -1)))
```

Comment:
I'm not sure about this structure. It seems simpler to have an inner function that removes the last stage, empty or not (or the specified stage I guess) and another that calls it in a `loop` until `has-clauses?` is true for the last stage.